File: windows/tests/Elevate.Core.Tests/Managed/ManagedPolicyTests.cs
using Elevate.Core.Managed;
using Elevate.Core.Models;
using FluentAssertions;

namespace Elevate.Core.Tests.Managed;

public class ManagedPolicyTests
{
    private static ManagedConfiguration Allowing(IReadOnlySet<SignInMethodKind>? kinds)
        => new() { AllowedSignInMethods = kinds };

    [Fact]
    public void NoAllowListAllowsEveryMethod()
    {
        var config = Allowing(null);

        foreach (var method in SignInMethod.BuiltIn.Append(SignInMethod.Custom("abc")))
        {
            ManagedPolicy.IsAllowed(method, config).Should().BeTrue();
        }
    }

    [Fact]
    public void AllowListRefusesMethodsOutsideIt()
    {
        var config = Allowing(new HashSet<SignInMethodKind> { SignInMethodKind.OwnApp });

        ManagedPolicy.IsAllowed(SignInMethod.OwnApp, config).Should().BeTrue();
        ManagedPolicy.IsAllowed(SignInMethod.AzureCLI, config).Should().BeFalse();
        ManagedPolicy.IsAllowed(SignInMethod.AzurePowerShell, config).Should().BeFalse();
        ManagedPolicy.IsAllowed(SignInMethod.Custom("abc"), config).Should().BeFalse();
    }

    [Fact]
    public void CustomInTheListAllowsAnyClientId()
    {
        var config = Allowing(new HashSet<SignInMethodKind> { SignInMethodKind.Custom });

        ManagedPolicy.IsAllowed(SignInMethod.Custom("abc"), config).Should().BeTrue();
        ManagedPolicy.IsAllowed(SignInMethod.Custom("11111111-2222-3333-4444-555555555555"), config).Should().BeTrue();
        ManagedPolicy.IsAllowed(SignInMethod.OwnApp, config).Should().BeFalse();
    }

    [Fact]
    public void EmptyAllowListAllowsNothing()
        => ManagedPolicy.IsAllowed(SignInMethod.OwnApp, Allowing(new HashSet<SignInMethodKind>())).Should().BeFalse();

    [Fact]
    public void NoTenantAllowListAllowsEveryTenant()
        => ManagedPolicy.IsTenantAllowed("11111111-2222-3333-4444-555555555555", null).Should().BeTrue();

    [Fact]
    public void TenantMatchIsCaseInsensitive()
    {
        var allowed = new HashSet<string> { "11111111-2222-3333-4444-AAAAAAAAAAAA" };

        ManagedPolicy.IsTenantAllowed("11111111-2222-3333-4444-aaaaaaaaaaaa", allowed).Should().BeTrue();
        ManagedPolicy.IsTenantAllowed("11111111-2222-3333-4444-AAAAAAAAAAAA", allowed).Should().BeTrue();
    }

    [Fact]
    public void TenantOutsideTheListIsRefused()
    {
        ManagedPolicy.IsTenantAllowed("t-other", new HashSet<string> { "t-home" }).Should().BeFalse();
        ManagedPolicy.IsTenantAllowed("t-home", new HashSet<string>()).Should().BeFalse();
    }
}

File: macos/Tests/ElevateCoreTests/ManagedPolicyTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct ManagedPolicyTests {
    private func config(allowing kinds: Set<SignInMethodKind>?) -> ManagedConfiguration {
        var config = ManagedConfiguration()
        config.allowedSignInMethods = kinds
        return config
    }

    @Test func noAllowListAllowsEveryMethod() {
        let config = config(allowing: nil)
        for method in SignInMethod.builtIn + [.custom(clientId: "abc")] {
            #expect(ManagedPolicy.isAllowed(method, by: config))
        }
    }

    @Test func allowListRefusesMethodsOutsideIt() {
        let config = config(allowing: [.ownApp])
        #expect(ManagedPolicy.isAllowed(.ownApp, by: config))
        #expect(!ManagedPolicy.isAllowed(.azureCLI, by: config))
        #expect(!ManagedPolicy.isAllowed(.azurePowerShell, by: config))
        #expect(!ManagedPolicy.isAllowed(.custom(clientId: "abc"), by: config))
    }

    @Test func customInTheListAllowsAnyClientId() {
        let config = config(allowing: [.custom])
        #expect(ManagedPolicy.isAllowed(.custom(clientId: "abc"), by: config))
        #expect(ManagedPolicy.isAllowed(.custom(clientId: "11111111-2222-3333-4444-555555555555"), by: config))
        #expect(!ManagedPolicy.isAllowed(.ownApp, by: config))
    }

    @Test func emptyAllowListAllowsNothing() {
        #expect(!ManagedPolicy.isAllowed(.ownApp, by: config(allowing: [])))
    }

    @Test func noTenantAllowListAllowsEveryTenant() {
        #expect(ManagedPolicy.isTenantAllowed("11111111-2222-3333-4444-555555555555", allowedIds: nil))
    }

    @Test func tenantMatchIsCaseInsensitive() {
        let allowed: Set<String> = ["11111111-2222-3333-4444-AAAAAAAAAAAA"]
        #expect(ManagedPolicy.isTenantAllowed("11111111-2222-3333-4444-aaaaaaaaaaaa", allowedIds: allowed))
        #expect(ManagedPolicy.isTenantAllowed("11111111-2222-3333-4444-AAAAAAAAAAAA", allowedIds: allowed))
    }

    @Test func tenantOutsideTheListIsRefused() {
        #expect(!ManagedPolicy.isTenantAllowed("t-other", allowedIds: ["t-home"]))
        #expect(!ManagedPolicy.isTenantAllowed("t-home", allowedIds: []))
    }
}
